### PR TITLE
]rating pallet - version 1

### PR DESCRIPTION
**Pull request by amarts**
* Testing in progress
* Yet to get identifier and arguments, methods finalised
* Will keep this as a reference PR for the progress of the work.

### CODE REVIEW

File: pallets/rating/src/ratings.rs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	/// Rating hash.
+	pub digest: HashOf<T>,
+	/// Rating controller.
+	pub controller: CordAccountOf<T>,
+	/// Rating (aggregated rating)
+	pub rating: u32,
+	/// Rating Count (Number of people rated)
+	pub count: u32,
+	/// Rating of identifier
+	pub entity: RatingEntityOf,
```
Comment (amarts) [inline pallets/rating/src/ratings.rs:51]: Also add 'sellerapp' and 'buyerapp' identifier as fields here

File: pallets/rating/src/ratings.rs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+#[derive(Clone, Encode, Decode, PartialEq, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+#[codec(mel_bound())]
+pub struct RatingType<T: Config> {
+	/// Rating hash.
+	pub digest: HashOf<T>,
+	/// Rating controller.
+	pub controller: CordAccountOf<T>,
+	/// Rating (aggregated rating)
+	pub rating: u32,
```
Comment (smohan-dw) [inline pallets/rating/src/ratings.rs]: should be a struct with ratings and count

File: pallets/rating/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+	#[pallet::storage_prefix = "Identifiers"]
+	pub type Ratings<T> =
+		StorageMap<_, Blake2_128Concat, IdentifierOf, RatingDetails<T>, OptionQuery>;
+
+	/// rating hashes stored on chain.
+	/// It maps from a rating hash to an identifier (resolve from hash).
+	#[pallet::storage]
+	#[pallet::storage_prefix = "Hashes"]
+	pub type RatingHashes<T> =
+		StorageMap<_, Blake2_128Concat, HashOf<T>, IdentifierOf, OptionQuery>;
```
Comment (amarts) [inline pallets/rating/src/lib.rs:134]: we also need query on 'entity'. 

Also we need an array which can keep a journal of all the transaction.

Comment (amarts) [reply pallets/rating/src/lib.rs:134]: There should be a transaction-id which is unique for rating.

File: pallets/rating/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+			Self::deposit_event(Event::Anchor {
+				identifier,
+				digest: tx_rating.digest,
+				author: tx_rating.controller,
+			});
+
+			Ok(())
+		}
+	}
```
Comment (amarts) [inline pallets/rating/src/lib.rs:249]: Should we be having 'update()' method here? Depends on what updates would happen to rating.

File: pallets/rating/src/ratings.rs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	}
+}
+
+/// An on-chain schema details mapped to an identifier.
+#[derive(Clone, Encode, Decode, PartialEq, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+#[codec(mel_bound())]
+pub struct RatingParams<T: Config> {
+	/// Rating identifier
+	pub identifier: IdentifierOf,
```
Comment (amarts) [inline pallets/rating/src/ratings.rs:87]: even here we may need 'entity'

File: pallets/rating/src/weights.rs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+/// Weight functions needed for pallet_rating.
+pub trait WeightInfo {
+	fn create() -> Weight;
+	fn update() -> Weight;
+	fn revoke() -> Weight;
+	fn remove() -> Weight;
+	fn council_remove() -> Weight;
+	fn digest() -> Weight;
+	fn delegate() -> Weight;
+	fn undelegate() -> Weight;
```
Comment (amarts) [inline pallets/rating/src/weights.rs]: remove these as its not present as method in lib.rs

File: pallets/rating/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+	pub struct Pallet<T>(_);
+
+	/// rating identifiers stored on chain.
+	/// It maps from an identifier to its details. Chain will only store the
+	/// last updated state of the data.
+	#[pallet::storage]
+	#[pallet::storage_prefix = "Identifiers"]
+	pub type Ratings<T> =
+		StorageMap<_, Blake2_128Concat, IdentifierOf, RatingDetails<T>, OptionQuery>;
+
```
Comment (smohan-dw) [inline pallets/rating/src/lib.rs:128]: Aggregated storage mapping entity identifier to ratings (overall, delivery)

File: pallets/rating/src/lib.rs (hunk 3 above)
Comment (amarts) [reply pallets/rating/src/lib.rs:134]: In progress. Good for version 1... soon will get this working.

File: pallets/rating/src/lib.rs (hunk 4 above)
Comment (amarts) [reply pallets/rating/src/lib.rs:249]: no update for now

File: pallets/rating/src/ratings.rs (hunk 5 above)
Comment (amarts) [reply pallets/rating/src/ratings.rs:87]: Done

**Pull request merged**